File: 0101-symmetric-tree/0101-symmetric-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool subTree(TreeNode* left_root, TreeNode* right_root){
        if(left_root == nullptr && right_root == nullptr) return true;
        if(left_root == nullptr || right_root == nullptr) return false;
        if(left_root -> val != right_root -> val) return false;
        
        return subTree(left_root -> left, right_root -> right) && subTree(left_root -> right, right_root -> left);     
    }
    bool isSymmetric(TreeNode* root) {
        if(root -> left == nullptr && root -> right == nullptr) return true;
        else if(root -> left == nullptr || root -> right == nullptr) return false;
        return subTree(root -> left, root -> right);
    }
};

File: 0101-symmetric-tree/0101-symmetric-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private boolean subTree(TreeNode left_root, TreeNode right_root){
        if(left_root == null && right_root == null) return true;
        if(left_root == null || right_root == null) return false;
        if(left_root.val != right_root.val) return false;
        
        return subTree(left_root.left, right_root.right) && subTree(left_root.right, right_root.left);     
    }
    public boolean isSymmetric(TreeNode root) {
        if(root.left == null && root.right == null) return true;
        else if(root.left == null || root.right == null) return false;
        return subTree(root.left, root.right);
    }
}